feat(JCE-658): Extension Providers (#700)
mia-platform/documentation: docs/console/company-configuration/providers/extensions/orchestrator-generator.md

---
id: orchestrator-generator
title: Orchestrator Generator
sidebar_label: Orchestrator Generator
---

> Capability: `Orchestrator Generator`

Reviewer: how does this line appear in the page? 🤔

Author: removed this line, I have specified the capability name in the text

> An Orchestrator Generator can generate orchestrator's **files**, ready to **deploy**, which are saved in the **configuration repository**. For instance, it could generate Kubernetes files.

Reviewer: ```suggestion
An Orchestrator Generator can generate orchestrator's **files**, ready to be **deployed**; these files are saved in the **configuration repository**. For instance, it could generate Kubernetes files.
```

If the Extension is enabled, the Console will send a **POST request** to its Provider's API Base URL to generate the deploy files. The request body will contain some information about the Project and the Configuration. The Console will expect to receive, in response, the files.

## Request format

### Headers

> For security concerns, the request is provided with a specific header, `X-Mia-Signature`, which can be used to validate that the payload comes from the Console.

Reviewer: this has not been implemented, the api is done using the authentication method specified in the provider configuration. (bearer token or basic auth)

Author: removed, now I talk about authentication

---
id: orchestrator-generator
title: Orchestrator Generator
sidebar_label: Orchestrator Generator
---

![Orchestrator Generator](img/orchestrator-generator.png)

:::info

An Orchestrator Generator can only be used when the [Enhanced Project Workflow](/development_suite/set-up-infrastructure/enhanced-project-workflow.md) is enabled.

:::

> One of the main features of Mia-Platform is the automatic generation of the **orchestrator's files**, ready to **be deployed**; these files are saved in the **configuration repository**. For instance, the Console could generate Kubernetes files.

Reviewer: ```suggestion
One of the main features of Mia-Platform is the automatic generation of the **orchestrator's files**, ready to **be deployed**; these files are saved in the **configuration repository**. For instance, the Console can generate Kubernetes files and deploy them on your runtime based on the Design configuration you provided.
```

maybe link the Design word to the proper page?

Author: done, with the link

It could be that you would want to generate those files differently: if this is the case, you should use an **Orchestrator Generator**.

> When the Orchestrator Generator Extension is enabled, the Console sends a **POST request** to its Provider's API Base URL to generate the deploy files. The request body contains some information about the Project and the Configuration, and the Console expects to receive, in response, the files.

Reviewer: ```suggestion
When a Provider with the Orchestrator Generator capability is enabled, the Console sends a **POST request** to the Provider API Base URL to generate the deploy files. The request body contains some information about the Project and the Configuration, and the Console expects to receive, in response, the files to be saved and deployed.
```

Author: is the Provider enabled, or is the Extension enabled by selecting a Provider?

Reviewer: interesting point; let's keep the way you wrote

## Setup

To enable the Orchestrator Generator on a Project, follow these steps:

- check that the [Enhanced Project Workflow](/development_suite/set-up-infrastructure/enhanced-project-workflow.md) is enabled for the Project;
- connect a Provider of type Extension API Provider with the **Orchestrator Generator capability** to its Company, as described in the [Extension API Provider page](/console/company-configuration/providers/extension-providers.md);
- navigate to [Project Settings](/console/project-configuration/project-settings.md) and click on the _Providers_ tab;
- enable the Orchestrator Generator extension in the _Extensions_ section, selecting the Provider previously connected.

![Extension Provider banner](img/orchestrator-generator-activation.png)

The Providers suitable for the Orchestrator Generator Extension are the ones featuring the **Orchestrator Generator** capability.
Every Provider with such capability can be selected to enable the Extension.

If no Provider with the _Orchestrator Generator_ capability is connected to the Company, **the Orchestrator Generator Extension will not appear in Project Settings**.

## Sequence diagram

> This is a simplified schema of the Orchestrator Generator's functioning.

Reviewer: is "functioning" a noun that can be used with the possessive `'s` ? 🤔

Author: yes: https://dictionary.cambridge.org/dictionary/english/functioning

but, if you prefer, I could use "operation"

Author: used "operation"

> ![Orchestrator Generator Sequence diagram](img/orchestrator-generator-sequence-diagram.png)

Reviewer: declare the sequence in place as code rather then include the png

| `periodSeconds` | Number | Value for the `periodSeconds` k8s configuration  | ✔️ | 
| `timeoutSeconds` | Number | Value for the `timeoutSeconds` k8s configuration  | ✔️ | 
| `successThreshold` | Number | Value for the `successThreshold` k8s configuration  | ✔️ | 
| `failureThreshold` | Number | Value for the `failureThreshold` k8s configuration  | ✔️ |

#### ConfigMount Schema

Each `configMaps` item holds information about how to mount a specific ConfigMap found in the top-level `configMaps` field.

Mount information is the following:

| Field   | Type     | Description     | Optional |
|---------|----------|-----------------|----------|
| `name` | String | Name of the ConfigMap to be mounted | |
| `mountPath` | String | Path where the configuration files should be mounted | |
| `subPaths` | String[] | List of specific files that needs to be picked from the ConfigMap and mounted | ✔️ |

#### Container schema

Each additional container has a similar shape of the main service, holding the following properties subset: `name`, `dockerImage`, `exclusiveServiceExposure`, `containerPorts`, `probes`, `environment`, `resources`, and `configMaps`.

> #### Service data model example

Reviewer: put the example in a collapsed item so that it doesn't clutter the page

Author: done
